README
(G)WOCSS - A diagnostic windfield model for complex terrain.
  Original code: Copyright (C) 2002 Francis Ludwig
  GNU/Linux port: Copyright (C) 2003-2015 Stephen L Arnold

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

This is the GPL release of Winds On Critical Streamline Surfaces (GWOCSS), 
a diagnostic windfield model for complex terrain.  It can be used for both 
objective analysis (with physics and terrain) of observational data and as 
a way to downscale prognostic model output (ie, to obtain finer resolution 
winds from a coarser grid model output).

BUILDING GWOCSS:

Consult the INSTALL file to learn how to build and install binaries of 
GWOCSS.  

You can override the usual user flags for debugging and/or optimization;
the configure flag --enable-debug is currently unused, so you will need
to override CC and CCC with configure to change debug/optimization flags.

It should only require a basic gfotran toolchain, ie, autotools and gcc built
with FORTRAN support.

The FFLAGS are specific to gfortran, but the basic autoconf macros allow for 
several other Fortran compilers.  You can override the FFLAGS on the command 
line or edit src/Makefile.in if you're not using GNU Fortran.  Use the model data 
and configuration included with the source to verify correct operation.  The
current code defaults to running in the parent directory of the input/output 
directories.

Note: previous versions required (static) libm.a or the glibc-static package
from your distro.  This build should produce a shared executable with a
chunk of static lilbgfotran bundled in.

To build and test gwocss manually, clone the repo or download the tarball.
Unpack and cd into the top-level gwocss src directory and run:

$ ./autogen.sh
$ ./configure (autogen will run configure by default)
$ make
$ gwocss  (this will use the sample data)

USING GWOCSS:

The Gentoo package will install some helper scripts and a config file to setup
the gwocss sample data and runtime directories based on the domain(s) in the
configuration file (the default is the "demo" domain which uses the sammple
data from Salt Lake).  Add yourself to the gwocss group and then add your
own domains in the same way, eg, make a direectory for your domain with 
sclin and slcout subdirs.  Assuming you already have config, topography,
and the minimum input data:

 $ mkdir -p /var/lib/gwocss/test1/{slcin,slcout}
 $ cp ${HOME}/test1/input_files/* /var/lib/gwocss/test1/slcin/
 $ gwocss test1

Alternatively, you can run the gwocss binary from the build directory in your
$HOME dir somewhere once you have your own domain setup.  For production use
you would need something to retrieve/prepare the input files and something
else to trigger the model when new input data is available (some of these
tools are on the TODO list, feel free to contribute ;)

The basic setup uses the slcin and slcout directories for the default input
and output files; legacy Fortran directory names have been converted to
lowercase, but not the filenames:

slcin/RUNSTF1.DAT - runtime configuration and localization parameters,
                    with sample data defaults (location near Salt Lake
                    City, UT).

slcin/SLC1KM.DAT - Topographic grid (1 km spacing) for above location.

slcin/SLCFILES - names of input files (meteorological data, only one file
                 for the sample problem).  See the next file...

slcin/10162215WXIN - Input meteorological data, can contain both surface
                     stations and upper air profiles.

inc/NGRIDS.PAR - Grid parameters and filename & path sizes.

inc/* - Fortran common include files.

The file GWOCSS_overview.pdf contains details on the sample problem, as
well as the format and naming convention for the input meteorological data
file(s).  Note the changes mentioned at the top of the document; the files
provided for the sample problem are the current (correct) format (you can
also read the source code).  Also, please ignore the Matlab references...

Please contact Steve Arnold <stephen.arnold42 _at_ gmail.com> for any questions
concerning this release.

The following recent AMS slides/abstract give a brief history/overview of GWOCSS:

http://tinyurl.com/GWOCSS-intro

See the following references for model details:

Ludwig, F. L., J. M. Livingston, and R. M. Endlich, 1991: "Use of Mass
   Conservation and Dividing Streamline Concepts for Efficient Objective
   Analysis of Winds in Complex Terrain," J. Appl. Meteorol., 30, 1490-1499

Ludwig, F. L. and D. Sinton, 2000; Evaluating an Objective Wind Analysis
   Technique with a Long Record of Routinely Collected Data, J. Appl.
   Meteorol., 39, 335-348.

Ludwig, F. L. and R. L. Street, 1995; Modification of Multiresolution Feature
   Analysis for Application to Three-Dimensional Atmospheric Wind Fields, J.
   Atmos. Sci., 52, 139-157.

Ludwig, F. L., R. L. Street, J. M. Schneider and K. R. Costigan, 1996:
   Analysis of Small-Scale Patterns of Atmospheric Motion in a Sheared,
   Convective Boundary Layer, J. Geophys. Res. (Atmospheres), 101D,
   9391-9411.



NEWS
2015-09-22

2.2.4_pre20150922 prerelease

Updated autotools build config, compiler options, etc,   Added warning flags
and IEEE math flags, started verification of floating point implementation
and refactoring plan.  Gentoo package and documentation updates on-going.

2013-11-26

Extracted from subversion and migrated to github.  No changes were made for
this migration except adding the PDF document and minor updates to this file
and the README.

2009-04-29

Updated build configuration for newer autotools and gcc.

2005-01-30

Initial release of GNU-WOCSS prepared with basic configure script and 
automake config that attempts to check for several Fortran compilers 
and environments.  Upgrades and other contributions are welcome...

Please contact Steve Arnold <stephen.arnold42 _at_ gmail.com> for any questions 
concerning these releases.


src/wocss.f
C
C********************************************************************
        PROGRAM  WOCSS
C********************************************************************
C
C       WINDS ON CRITICAL STREAMLINE SURFACES (WOCSS) PROGRAM FINDS 
C       TOPOGRAPHICALLY INDUCED WINDS BY MAKING THE ORIGINALLY ANALYZED 
C       WINDS NONDIVERGENT WITHIN FLOW SURFACES THAT ARE DEFINED USING 
C       A CONCEPT ANALAGOUS TO THE CRITICAL STREAMLINE WHEN THE 
C       ATMOSPHERE IS STABLY STRATIFIED. FLOW SURFACES CAN INTERSECT 
C       THETERRAIN RESULTING IN ZERO WINDS IN THESE 'UNDERGROUND CELLS.' 
C       ADJUSTMENT TOWARD NONDIVERGENCE IN BAL5 CAUSES FLOW AROUND THE 
C       OBSTACLES.  GEOSIG READS IN WIND SOUNDINGS AND SURFACE DATA.  
C       DIRECT VECTOR ALTERATIONS ARE USED IN SUBROUTINE BAL5. WIND 
C       COMPONENTS AT SPECIFIED POINTS CAN BE HELD CONSTANT OR ADJUSTED 
C       AT A FRACTION OF THE ADJUSTMENTS IN OTHER CELLS IN BAL5.
C
C       THE MODEL HAS BEEN DESCRIBED IN:
C                LUDWIG, F. L., J. M. LIVINGSTON, AND R. M. ENDLICH, 
C                1991:  'USE OF MASS CONSERVATION AND DIVIDING 
C                STREAMLINE CONCEPTS FOR EFFICIENT OBJECTIVE ANALYSIS 
C                OF WINDS IN COMPLEX TERRAIN,' 
C                  J. APPL. METEOROL., VOL. 30, PP. 1490-1499.
C
C       THIS VERSION HAS REMOVED SEVERAL OPTIONS FOUND IN SOME EARLIER 
C       VERSIONS.  ONLY ONE GRID IS USED, NO NESTING.   AT LEAST ONE 
C       SOUNDING IS REQUIRED; THE GEOSTROPHIC WIND PROVISION IS NO 
C       LONGER AVAILABLE.  THE LOW ALTITUDE GRID POINTS TO BE USED
C       FOR DEFINING CRITICAL STREAMLINE WINDS ARE NOW INPUT.  
C
C       NEW PROVISIONS INCLUDE A PARAMETER INPUT FILE SO THAT 
C       ADJUSTMENT FACTORS, FLOW SURFACE COMPRESSION AND OTHER FACTORS 
C       CAN BE CHANGED WITHOUT RECOMPILING.  THE NUMBER OF LABELED 
C       COMMONS HAS ALSO BEEN SUBSTATIALLY REDUCED.
C
C       JANUARY 2002   
C                    F. LUDWIG
C                    ENVIRONMENTAL FLUID MECHANICS LAB
C                    DEPT. OF CIVIL ENGINEERING
C                    STANFORD UNIVERSITY
C                    STANFORD, CA 94305-4020
C
C			BASED IN LARGE PART ON EARLIER WORK WITH:
C			R. ENDLICH, A. BECKER, D. SINTON, K. NITZ, 
C                     J. LIVINGSTON, B. MORLEY AND C. BHUMRALKAR
C
C*********************************************************************
C  ADJMAX           THE FRACTION OF THE USUAL ITERATIVE ADJUSTMENT 
C                   TOWARD NONDIVERGENCE THAT IS MADE AT GRID POINTS 
C                    NEAR OBSERVATIONS IN SUBROUTINE BAL5.
C  AVTHK            HEIGHT AGL OF TOP SURFACE OVER LOWEST TERRAIN.
C  CMPRES           MAXIMUM 'COMPRESSION' OF SURFACES -- 0 MEANS THAT  
C                   LOWER SFCS MUST PARALLEL THOSE ABOVE, 1 MEANS THAT  
C                   THE LOW SFC CAN TOUCH THE NEXT LEVEL -- DEFINES  
C                   INFLUENCE OF UPPER STABLE LAYERS ON LESS STABLE  
C                   LOWER ONES.
C  DBUG,DBUG2       LOGICAL FLAGS TO TRIGGER PRINTING OF STATEMENTS IN  
C                   PROGRAM
C  D2MIN            THE MINIMUM DIDTANCE ALLOWED IN THE INVERSE  
C                   WEIGHTING DENOMINATOR.
C  DPOTMIN	    THE MINIMUM ALLOWED POTENT TEMP LAPSE RATE  
C                   (DEG/M) -- LIMITS INSTABILITY AND HENCE RISE OF  
C                   FLOW SFCS.
C  DS               GRID SIZE IN METERS 
C  DSCRS            GRID SIZE (KM)
C  DTWT             DISTANCE WEIGHT POWER --WEIGHT=1/(DISTANCE**DTWT)
C  DZMAX(JT,JZ)     MAXIMUM RISE FOR JZTH FLOW SFC AS DETERMINED FROM  
C                   T-SONDE JT 
C  GRDHI            HEIGHT MSL OF HIGHEST TERRAIN ON GRID.
C  MDATE            DAY OF MONTH
C  KGRIDX           X INDEX OF REFERENCE POINT           
C  KGRIDY           Y INDEX OF REFERENCE POINT
C  LOWIX(JL)        X INDICES OF LOWEST PTS 
C  LOWIY(JL)        Y INDICES OF LOWEST PTS 
C  NCOL             NUMBER OF COLUMNS (X INDEX) IN GRID             
C  NCOLM1           NCOL-1
C  NEND             NUMBER OF HOURS TO CALCULATE BEFORE QUITTING IN A  
C                   SEQUENCE OF DATA)
C  NLVL             NUMBER OF VERTICAL FLOW SURFACES
C  NROW             NUMBER OF ROWS (Y INDEX) IN GRID
C  NROWM1           NROW-1
C  NSITES   	    MAXIMUM NUMBER OF SITES (INCLUDING UPPER AIR) TO 
C                   BE USED.
C  NSNDHT   	    MAXIMUM NUMBER SOUNDING (WIND OR TEMP) LEVELS TO  
C                   BE USED.
C  NTSITE   	    MAXIMUM NUMBER T-SONDES TO BE USED.
C  NUMDOP           NUMBER OF UPPER WIND SITES
C  NUMNWS           NUMBER OF SURFACE WIND SITES
C  NUMTMP           NUMBER OF TEMEPRATURE SOUNDING SITES
C  NUMTOT           TOTAL NUMBER OF OBSERVING WIND SITES
C  NWSITE   	    MAXIMUM NUMBER WIND SOUNDINGS TO BE USED.
C  NXGRD    	    DIMENSION FOR GRIDS IN W-E DIRECTION.
C  NYGRD    	    DIMENSION FOR GRIDS IN S-N DIRECTION.
C  NZGRD    	    DIMENSION FOR FLOW SURFACES.
C  PWR		    POWER TO WHICH SEPARATION IS RAISED FOR INVERSE  
C                   DISTANCE WTS.
C  RHS(JX,JY,JZ)    HEIGHT (M) ABOVE TERRAIN OF JZTH FLOW SFC ABOVE  
C                   PT JX,JY
C  RHSLO(JT,JZ)     HEIGHT (M) OF JZTH FLOW SFC ABOVE THE LOWEST  
C                   TERRAIN NEAR T-SONDE JT  
C  SFCHT(JX,JY)     TERRAIN HEIGHT (M MSL) AT PT JX,JY
C  SFCLOW           TERRAIN HEIGHT (M MSL) OF LOWEST PT IN THE DOMAIN
C  SIGMA(JZ)        FRACTION (OVER LOWEST TERRAIN) OF THE HEIGHT OF 
C                   THE TOP SURFACE FOR SFC JZ
C  SLFAC            CONTROLS THE DEGREE TO WHICH THE 1ST GUESS SURFACES  
C                   FOLLOW THE TERRAIN -- 0 GIVES FLAT SFCS & 1 GIVES  
C                   TERRAIN-FOLLOWING SFCS.
C  SPDCNV           CONVERSION FACTOR -- INPUT UNITS TO M/S
C  U(JX,JY,JZ)      WESTERLY COMPONENT (M/S) AT (JX,JY,JZ)
C  UCOMP(JOB)       OBSERVED WESTERLY COMPONENT (M/S) AT SITE JOB
C  USIG(JWS,JZ)     U COMPONENT AT FLOW SFC JZ ON WIND SOUNDING JWS
C  UTMAPX           UTM EASTING COORDINATE (KM) OF REFERENCE PT  
C                   (KGRIDX,KGRIDY).
C  UTMAPY           UTM NORTHING COORDINATE (KM) OF REFERENCE PT  
C                   (KGRIDX,KGRIDY).
C  V(JX,JY,JZ)      SOUTHERLY COMPONENT (M/S) AT (JX,JY,JZ)
C  VCOMP(JOB)        OBSERVED SOUTHERLY COMPONENT (M/S) AT SITE JOB
C  VSIG(JWS,JZ)     V COMPONENT AT FLOW SFC JZ ON WIND SOUNDING JWS
C  W(JX,JY,JZ)      UPWARD COMPONENT (M/S) AT (JX,JY,JZ)
C  XG(JSITE)        COORDINATE (GRID UNITS) OF OBSERVATION SITE JSITE.
C  XORIG            UTM EASTING COORDINATE (KM) OF THE GRID ORIGIN. 
C  YG(JSITE)        COORDINATE (GRID UNITS) OF OBSERVATION SITE JSITE.
C  YORIG            UTM NORTHING COORDINATE (KM) OF THE GRID ORIGIN.
C                       (NOTE: XORIG,YORIG ARE AT 1,1)
C  ZRISE            DIFFERENCE BETWEEN SFCHI -SFCLOW.
C  ZZERO            ROUGHNESS LENGTH (M) FOR LOG PROFILE ESTIMATES.    
C
	INCLUDE 'NGRIDS.PAR'
C
	INCLUDE 'ANCHOR.CMM'
	INCLUDE 'FLOWER.CMM'
	INCLUDE 'LIMITS.CMM'
	INCLUDE 'STALOC.CMM'
C
        CHARACTER*1 CHGRID
        CHARACTER*6 CHDDHR
	CHARACTER*12 CHFNAM
	CHARACTER*8 CHMMDDTTTT
C
	INTEGER*4 NREAD,KEY
C
        DATA KEY /16/
C
        DATA RHSLO /NARRAY*0.0/
C
C!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
C  MAKE SURE TO SET PATH DEFINITION APPROPRIATE TO YOUR JOB
C  STRING LENGTH FOR CHPATH IS DEFINED IN INCLUDE FILE, NGRIDS.PAR'
C!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
C
	CHPATH='slcin/'
C
C  GRID SIZE SELECTION OPTION DISABLED BY ASSIGNING A 1 KM MESH SIZE
C
	CHMESH='1'
C
C       SET PATH & OPEN INPUT PARAMETER FILE
C
        CALL RUNRD (NIT,KFYL1,KFYL2)
	NREAD=0
C
C ****************DISABLED******************************
C  OPEN FILES FOR KEEPING TRACK OF WHAT HAPPENED WHEN & TEST OUTPUT.
C
C 	OPEN(33,FILE='CHPOUT//'TEST'//CHMESH//'KM',
C     $	     FORM='FORMATTED', STATUS='UNKNOWN')
C ****************DISABLED******************************
C
	WRITE (*,*) 'STARTING'
C
C    OPEN FILE 11 (TOPOGRAPHY)
C    ASSUMED SQUARE GRID, DEFAULT 80 x 80 FOR VTMX CASE
C
	LGRD=NINT(10.0*DSCRS)
	IF (LGRD.EQ.10 .OR. LGRD.EQ.20 .OR. LGRD.EQ.50) THEN
	   CHGRID=CHAR(NINT(48+DSCRS))
	ELSE
	   CHGRID='H'
	END IF
C
	WRITE (*,*)'OPENING '//CHPATH//'SLC1KM.DAT'
C
       	OPEN (11,FILE=CHPATH//'SLC1KM.DAT',STATUS='OLD',
     $         FORM='FORMATTED')
C
C  READ TERRAIN HEIGHTS FOR ALL GRIDS USING TOPO
C
	CALL TOPO
        DS=DSCRS*1.0E3
        NCOLM1=NCOL-1
        NROWM1=NROW-1
C
	CLOSE (11)
C
       	OPEN(11,FILE=CHPATH//'SLCFILES',STATUS='OLD',
     $         FORM='FORMATTED')
C
	NREAD=0
C
44	READ (11,6003,END=8586) CHFNAM
	CHMMDDTTTT=CHFNAM(1:8)
C
	WRITE (*,*) 'OPENING '//CHPATH//CHFNAM
C
	OPEN (12,FILE=CHPATH//CHFNAM,STATUS='OLD',
     $         FORM='FORMATTED')
C
C* READ AND ANALYZE WIND DATA USING WXANAL
C* MAKE INITIAL WIND ANALYSIS ON MESH
C
	READ (12,6001,END=8686) IHOUR,IMIN,IMO,MDATE,IYEAR
C
	NREAD=NREAD+1
	IF (NREAD .GT. NEND)  GO TO 8586
C
	KHR=IHOUR
 	KMON=IMO
	KDAY=MDATE
	LASTMO=IMO
	MINUTZ=JULMIN(IYEAR,IMO,MDATE,IHOUR,IMIN)
C
C
C  GEOSIG READS METEOROLOGICAL INPUTS & CALLS ROUTINES FOR
C  INTERPOLATION AND DEFINING SURFACE SHAPES.
C
	CALL GEOSIG
C
C CALL SUBROUTINE TO MAKE WINDS NONDIVERGENT
C
	CALL BAL5(NIT)
C
C  IF NEEDED, GET PRESSURES AND TEMPERATURES OON FLOW SURFACES.
C
	IF (DOTHET .OR. DOPRES .OR. DOBRI .OR. DOBVPD) CALL BETPAR
C
C  GET PARAMETER VALUES ON FLAT SURFACES BY INTERPOLATIO IN Z.
C
        CALL TSTWND(CHMMDDTTTT)
C
C  IF REQUESTED, CALCULATE BULK RICHARDSON NUMBER AND BRUNT-VAISALA 
C  PERIOD ON PLOTTING SURFACES.
C
	IF (DOBRI .OR. DOBVPD) CALL GETRICH
C
	CHDDHR=CHMMDDTTTT(3:8)
C
C  WRITE OUTPUT FILES
C
	CALL PUTOUT (CHDDHR,CHFNAM,MINUTZ,NREAD)
C
	GO TO 44
C
8586    CONTINUE
C
8686	CONTINUE
C
	IF (NREAD .GT. NEND)  GO TO 8786
C
8786	IF (NREAD .GT. NEND) NREAD=NEND
	IF (DOWVSZ) CALL PUTOUT (CHDDHR,CHFNAM,MINUTZ,-NREAD)
	WRITE (*,*) 'FINISHED '
C
6001	FORMAT (4I2,1X,I4)	
6003	FORMAT (A12)
C
	STOP
C
	END

